client/state/notices/reducer.js

```diff
@@ -2,39 +2,103 @@
  * External dependencies
  */
 import { combineReducers } from 'redux';
+import omit from 'lodash/omit';
+import reduce from 'lodash/reduce';
+import { translate } from 'i18n-calypso';
 
 /**
  * Internal dependencies
  */
 import {
-	NEW_NOTICE,
-	REMOVE_NOTICE,
-	ROUTE_SET,
-	SERIALIZE,
-	DESERIALIZE
+	NOTICE_CREATE,
+	NOTICE_REMOVE,
+	POST_SAVE_SUCCESS,
+	ROUTE_SET
 } from 'state/action-types';
+import { createReducer } from 'state/utils';
 
-export function items( state = [], action ) {
-	switch ( action.type ) {
-		case NEW_NOTICE:
-			return [ action.notice, ...state ];
-		case REMOVE_NOTICE:
-			return state.filter( ( notice ) => ( notice.noticeId !== action.noticeId ) );
-		case ROUTE_SET:
-			return state.filter( notice => {
-				const show = notice.isPersistent || notice.displayOnNextPage;
-				if ( notice.displayOnNextPage ) {
-					notice.displayOnNextPage = false;
+export const items = createReducer( {}, {
+	[ NOTICE_CREATE ]: ( state, action ) => {
+		const { notice } = action;
+		return {
+			...state,
+			[ notice.noticeId ]: notice
+		};
+	},
+	[ NOTICE_REMOVE ]: ( state, action ) => {
+		const { noticeId } = action;
+		if ( ! state.hasOwnProperty( noticeId ) ) {
+			return state;
+		}
+
+		return omit( state, noticeId );
+	},
+	[ ROUTE_SET ]: ( state ) => {
+		return reduce( state, ( memo, notice, noticeId ) => {
+			if ( ! notice.isPersistent && ! notice.displayOnNextPage ) {
+				return memo;
+			}
+
+			let nextNotice = notice;
+			if ( nextNotice.displayOnNextPage ) {
+				nextNotice = {
+					...nextNotice,
+					displayOnNextPage: false
+				};
+			}
+
+			memo[ noticeId ] = nextNotice;
+			return memo;
+		}, {} );
+	},
+	[ POST_SAVE_SUCCESS ]: ( state, action ) => {
```

Comment:
>I like putting this in notices reducer instead of explicitly calling notice create.

Indeed, I like that it helps keeps notice behavior self-contained, having the notices being concerned themselves with actions going through the system, not other areas of the tree explicitly calling to create notices.

>I also was thinking about some notices middleware that would just catch metadata in any action.

I think this could be a good thing, or at least to have as an option. I'm not sure in which case we'd prefer to use the metadata than adding to the reducer here, though. Thoughts?

>As an aside, showing the notice on success feels sluggish. I wonder if we should be optimistic in displaying `X posts moved to trash`, and modify the notice content on error.

My initial thinking is that the sluggishness is intentional, reflecting the result of the network request. Do you think it might be a bit strange to see "successfully moved" notice, replaced by an error notice if the attempt failed? Alternatively, we could add a "Saving..." info notice, but that feels a bit clunky / excessively verbose.